core/src/main/java/org/kiwix/kiwixmobile/core/history/viewmodel/HistoryViewModel.kt

```diff
@@ -0,0 +1,217 @@
+      searchString,
+      showAllSwitchOn
+    )
+  }
+
+  private fun selectStateBasedOnInput(
+    currentBook: String,
+    historyListWithDateItems: List<HistoryListItem>,
+    selectedItems: List<HistoryListItem>,
+    searchString: String,
+    showAllSwitchOn: Boolean
+  ) = when {
+    historyListWithDateItems.isEmpty() -> NoResults(searchString, historyListWithDateItems)
+    selectedItems.isNotEmpty() -> SelectionResults(
+      searchString,
+      historyListWithDateItems,
+      selectedItems,
+      showAllSwitchOn,
+      currentBook
+    )
+    else -> Results(searchString, historyListWithDateItems, showAllSwitchOn, currentBook)
+  }
+
+  private fun addDateHeadersToHistoryItems(historyItemSearchResults: List<HistoryListItem>):
+    MutableList<HistoryListItem> {
+    return historyItemSearchResults.reversed().foldOverAddingHeaders(
+        { historyItem -> DateItem((historyItem as HistoryItem).dateString) },
+        { current, next -> (current as HistoryItem).dateString != (next as HistoryItem).dateString }
+      )
+  }
+
+  private fun filterOutAllSelectedHistoryItems(historyItemSearchResults: List<HistoryListItem>):
+    List<HistoryItem> {
+    return historyItemSearchResults.filterIsInstance<HistoryItem>().filter { it.isSelected }
+  }
+
+  private fun deselectAllHistoryItems(historyItemSearchResults: List<HistoryListItem>) {
+      historyItemSearchResults
+        .filterIsInstance<HistoryItem>()
+        .forEach { it.isSelected = false }
+  }
+
+  override fun onCleared() {
+    compositeDisposable.clear()
+    super.onCleared()
+  }
+
+  private fun actionMapper() = actions.map {
+    when (it) {
+      ExitHistory -> effects.offer(Finish)
+      is Filter -> filter.offer(it.searchTerm)
+      is ToggleShowHistoryFromAllBooks ->
+        toggleShowAllHistorySwitchAndSaveItsStateToPrefs(it.isChecked)
+      is CreatedWithIntent -> filter.offer(it.searchTerm)
+      is OnItemLongClick -> selectItemAndOpenSelectionMode(it.historyItem)
+      is OnItemClick -> appendItemToSelectionOrOpenIt(it)
+      is RequestDeleteAllHistoryItems ->
+        openDialogToRequestDeletionOfAllHistoryItems(it.dialogShower)
+      is RequestDeleteSelectedHistoryItems ->
+        openDialogToRequestDeletionOfSelectedHistoryItems(it.dialogShower)
+      ExitActionModeMenu -> deselectAllItems.offer(true)
+      DeleteHistoryItems -> effects.offer(DeleteSelectedOrAllHistoryItems(state, historyDao))
+    }
+  }.subscribe({}, Throwable::printStackTrace)
+
+  private fun toggleShowAllHistorySwitchAndSaveItsStateToPrefs(isChecked: Boolean) {
+    showAllSwitchToggle.offer(isChecked)
+    sharedPreferenceUtil.setShowHistoryCurrentBook(!isChecked)
+  }
+
+  private fun selectItemAndOpenSelectionMode(historyItem: HistoryItem) {
+    historyItem.isSelected = true
+    deselectAllItems.offer(false)
+  }
+
+  private fun openDialogToRequestDeletionOfSelectedHistoryItems(dialogShower: DialogShower) {
+    dialogShower.show(DeleteSelectedHistory, {
```

Comment:
This may want to be a SideEffect, strictly speaking the ViewModel shouldn't "do" "stuff"